wodles/azure/tests/test_azure.py

```diff
@@ -0,0 +1,859 @@
+        expected_str = f"TimeGenerated >= datetime({desired_date})"
+    else:
+        if parse(desired_date) < parse(min_date, fuzzy=True):
+            expected_str = f"( TimeGenerated < datetime({min_date}) and TimeGenerated >= datetime({desired_date})) " \
+                           f"or ( TimeGenerated > datetime({max_date}))"
+        elif parse(desired_date) > parse(max_date, fuzzy=True):
+            expected_str = f"TimeGenerated >= datetime({desired_date})"
+        else:
+            expected_str = f"TimeGenerated > datetime({max_date})"
+
+    assert expected_str in result["query"]
+
+
+@patch('azure-logs.logging.error')
+@patch('azure-logs.orm.get_row', side_effect=azure.orm.AzureORMError)
+def test_build_log_analytics_query_ko(mock_get, mock_logging):
+    """Test build_log_analytics_query handles ORM exceptions."""
+    with pytest.raises(SystemExit) as err:
+        azure.build_log_analytics_query(offset=None, md5_hash=None)
+    assert err.value.code == 1
+    mock_logging.assert_called_once()
+
+
+@pytest.mark.parametrize('file, time_position', [
+    ("log_analytics_events", 0),
+    ("log_analytics_events", None),
+    ("log_analytics_events_no_results", None),
+    ("log_analytics_events_empty", None)
+])
+@patch('azure-logs.logging.error')
+@patch('azure-logs.update_row_object')
+@patch('azure-logs.iter_log_analytics_events')
+@patch('azure-logs.get_time_position')
+@patch('azure-logs.get')
+def test_get_log_analytics_events(mock_get, mock_position, mock_iter, mock_update, mock_logging, file, time_position):
+    """Test get_log_analytics_events gets the logs, process the response and iterate the events if the 'TimeGenerated'
+    field was present."""
+    azure.args = MagicMock(la_query="test_query")
+    m = MagicMock(status_code=200)
+    with open(os.path.join(TEST_DATA_PATH, file)) as f:
+        try:
+            events = json.loads(f.read())
+            m.json.return_value = events
+            columns = events['tables'][0]['columns']
+            rows = events['tables'][0]['rows']
+        except (KeyError, json.decoder.JSONDecodeError):
+            m.json.return_value = {}
+            columns = None
+            rows = None
+
+    mock_get.return_value = m
+    mock_position.return_value = time_position
+    url = "url"
+    body = "body"
+    headers = "headers"
+    azure.get_log_analytics_events(url=url, body=body, headers=headers, md5_hash="")
+    mock_get.assert_called_with(url, params=body, headers=headers)
+    if rows is None or (len(rows) > 0 and time_position is None):
+        mock_logging.assert_called_once()
+    elif len(rows) == 0:
+        mock_position.assert_not_called()
+        mock_iter.assert_not_called()
+        mock_update.assert_not_called()
+    else:
+        mock_position.assert_called_with(columns)
+        mock_iter.assert_called_with(columns, rows)
+        mock_update.assert_called_once()
+
+
+@patch('azure-logs.get')
+def test_get_log_analytics_events_ko(mock_get):
```

Comment:
You are right, that particular test is not testing the `KeyError` exception. That's being tested in the previous test (`test_get_log_analytics_events`) using this block:

```
    if rows is None or (len(rows) > 0 and time_position is None):
        mock_logging.assert_called_once()
```

This line covers both `KeyError` exception and `No TimeGenerated field was found` error.


The test was splitted in this way because the `KeyError` test requires the same parameters that we were already using, while properly testing the `400` response from the request module requires to use different parametrize parameters.

Combining these tests or altering them would imply unnecessarily overcomplicating the parametrize we define for the different test cases.